bats/tests/k8s/up-downgrade-k8s.bats

```diff
@@ -0,0 +1,129 @@
+# Test case 8 & 13
+
+setup() {
+    load '../helpers/load'
+}
+
+@test 'factory reset' {
+    factory_reset
+}
+
+@test 'start rancher desktop' {
+    start_kubernetes
+    wait_for_apiserver
+    # the docker context "rancher-desktop" may not have been written
+    # even though the apiserver is already running
+    wait_for_container_engine
+}
+
+@test 'deploy nginx - always restart' {
+    ctrctl pull nginx
+    run ctrctl run -d -p 8585:80 --restart=always --name nginx-restart nginx
+    assert_success
+}
+
+@test 'deploy nginx - no restart' {
+    run ctrctl run -d -p 8686:80 --restart=no --name nginx-no-restart nginx
+    assert_success
+}
+
+@test 'deploy busybox' {
+    run kubectl create deploy busybox --image=busybox --replicas=2 -- /bin/sh -c "sleep inf"
+    assert_success
+}
+
+verify_nginx() {
+    for port in 8585 8686; do
+        run curl "http://localhost:$port"
+        assert_success
+        assert_output --partial "Welcome to nginx!"
+    done
+}
+
+@test 'verify nginx before upgrade' {
+    try verify_nginx
+}
+
+verify_busybox() {
+    run kubectl get pods --selector="app=busybox" -o jsonpath='{.items[*].status.phase}'
+    assert_output --partial "Running Running"
+}
+
+@test 'verify busybox before upgrade' {
+    try verify_busybox
+}
+
+verify_images() {
+    if using_docker; then
+        run docker images
+        assert_output --partial "nginx" "busybox"
+    else
+        run nerdctl images --format json
+        assert_output --partial '"Repository":"nginx'
+        run nerdctl --namespace k8s.io images
+        assert_output --partial "busybox"
+    fi
+}
+@test 'verify images before upgrade' {
+    verify_images
+}
+
+@test 'upgrade kubernetes' {
+    rdctl set --kubernetes.version "$RD_KUBERNETES_VERSION"
+    wait_for_apiserver "$RD_KUBERNETES_VERSION"
+    wait_for_container_engine
+}
+
+verify_nginx_after_change_k8s() {
+        run curl http://localhost:8686
+        assert_failure
+        assert_output --partial "Failed to connect to localhost port 8686"
+
+        run curl http://localhost:8585
+        assert_success
+        assert_output --partial "Welcome to nginx!"
+}
+
+@test 'verify nginx after upgrade' {
+    try verify_nginx_after_change_k8s
+}
+
+@test 'verify busybox after upgrade' {
+    try verify_busybox
+}
+
+@test 'verify images after upgrade' {
+    verify_images
+}
+
+@test 'restart nginx-no-restart before downgrade' {
+    run ctrctl start nginx-no-restart
+    try verify_nginx
```

Comment:
Every `run` and every `try` call should be followed by at least one `assert` or `refute` to check if the command succeeded or not.

I know we sometimes use `run ...` if we don't want a failure of the command to be treated as a test failure. It would probably be better to write this as `... || true` to show that we don't intend to capture `$output` or `$status`.

Also, as a special case instead of `run ...; assert_success` we can simply write `...`.

```suggestion
    ctrctl start nginx-no-restart
    try verify_nginx
    assert_success
```

With this modification we see that `nerdctl start` does actually fail:

```
✗ restart nginx-no-restart before downgrade
   (from function `nerdctl' in file tests/k8s/../helpers/commands.bash, line 70,
    from function `ctrctl' in file tests/k8s/../helpers/commands.bash, line 48,
    in test file tests/k8s/up-downgrade-k8s.bats, line 104)
     `ctrctl start nginx-no-restart' failed
   time="2023-03-21T05:28:22Z" level=fatal msg="1 errors:\nfailed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error running hook #0: error running hook: exit status 1, stdout: , stderr: time=\"2023-03-21T05:28:22Z\" level=fatal msg=\"failed to call cni.Setup: plugin type=\\\"bridge\\\" failed (add): failed to allocate for range 0: 10.4.0.3 has been allocated to default-67c1cac88ba6fe8bd3550eed187b8bf8084c6e5b0dbc24fae1885379b30795d5, duplicate allocation is not allowed\"\nFailed to write to log, write /var/lib/nerdctl/dbb19c5e/containers/default/67c1cac88ba6fe8bd3550eed187b8bf8084c6e5b0dbc24fae1885379b30795d5/oci-hook.createRuntime.log: file already closed: unknown"
   Error: exit status 1
```

And then `try verify_nginx` just continues to call `verify_nginx` until it runs out of retries. Without the `assert_success` behind it we would not notice this failure either.